dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java

```diff
@@ -1338,6 +1339,40 @@ public boolean handleHTTP2UpgradeRequest(Map<String, String> http2Settings) {
         VirtualConnection vc = link.getVirtualConnection();
         H2InboundLink h2Link = new H2InboundLink(channel, vc, getTCPConnectionContext());
 
+        if(this.isc != null) {
+            if(this.isc.isIncomingBodyExpected() && !this.isc.isBodyComplete()){
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc, "Body needed for request. Queueing data locally before upgrade.");
+                }
+                HttpInputStreamImpl body = this.request.getBody();
+                body.setupChannelMultiRead();
+                byte[] inBytes = new byte[1024];
+                try{
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Starting request read loop.");
+                    }
+                    for (int n; (n = body.read(inBytes)) != -1;) {}
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Finished request read loop.");
+                    }
+                }catch(Exception e){
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Got exception reading request and queueing up data. Can't handle request upgrade to HTTP2.", e);
+                    }
+                    vc.getStateMap().put(h2InitError, true);
```

Comment:
if there is an exception, how is the channel multi read buffer get cleanup?  It depends on the dataAlreadyReadFromChannel (true) (in the non custom property case) which now can't be set due to an exception.